src/Build/BuildCheck/Infrastructure/EditorConfig/EditorConfigParser.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Drawing.Design;
+using System.IO;
+using System.Linq;
+using System.Reflection.Metadata.Ecma335;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Build.Shared;
+using static Microsoft.Build.BuildCheck.Infrastructure.EditorConfig.EditorConfigGlobsMatcher;
+
+namespace Microsoft.Build.BuildCheck.Infrastructure.EditorConfig
+{
+    internal class EditorConfigParser
+    {
+        private const string EditorconfigFile = ".editorconfig";
+        private Dictionary<string, EditorConfigFile> editorConfigFileCache;
+
+        internal EditorConfigParser()
+        {
+            editorConfigFileCache = new Dictionary<string, EditorConfigFile>();
+        }
+
+        internal Dictionary<string, string> Parse(string filePath)
+        {
+            var editorConfigs = EditorConfigFileDiscovery(filePath);
+            return MergeEditorConfigFiles(editorConfigs, filePath);
+        }
+
+        /// <summary>
+        /// Fetches the list of EditorconfigFile ordered from the nearest to the filePath.
+        /// </summary>
+        /// <param name="filePath"></param>
+        internal IEnumerable<EditorConfigFile> EditorConfigFileDiscovery(string filePath)
+        {
+            var editorConfigDataFromFilesList = new List<EditorConfigFile>();
+
+            var directoryOfTheProject = Path.GetDirectoryName(filePath);
+            var editorConfigFilePath = FileUtilities.GetPathOfFileAbove(EditorconfigFile, directoryOfTheProject);
+
+            while (editorConfigFilePath != string.Empty)
+            {
+                EditorConfigFile editorConfig;
+
+                if (editorConfigFileCache.ContainsKey(editorConfigFilePath))
+                {
+                    editorConfig = editorConfigFileCache[editorConfigFilePath];
+                }
+                else
+                {
+                    var editorConfigfileContent = File.ReadAllText(editorConfigFilePath);
+                    editorConfig = EditorConfigFile.Parse(editorConfigfileContent);
+                    editorConfigFileCache[editorConfigFilePath] = editorConfig;
```

Comment:
Is it Ok for this class to not be thread safe?

Reply:
Discussed offline with @JanKrivanek and there is a case with graph enabled scenario where it could be used like that. So update it to the use ConcurrentDictionary.